libyul/optimiser/MemoryLoadResolver.cpp

```diff
@@ -0,0 +1,447 @@
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include "libyul/Dialect.h"
+#include "libyul/YulString.h"
+#include "libyul/optimiser/ASTWalker.h"
+#include "libyul/optimiser/CallGraphGenerator.h"
+#include "libyul/optimiser/DataFlowAnalyzer.h"
+#include "libyul/optimiser/NameDispenser.h"
+#include "libyul/optimiser/OptimiserStep.h"
+#include "libyul/optimiser/SSAValueTracker.h"
+#include <libyul/optimiser/MemoryLoadResolver.h>
+#include <libyul/optimiser/Semantics.h>
+
+#include <libyul/AST.h>
+#include <libyul/SideEffects.h>
+#include <libyul/Utilities.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/CommonData.h>
+
+#include <libsmtutil/SMTPortfolio.h>
+#include <libsmtutil/SolverInterface.h>
+#include <libsmtutil/Helpers.h>
+#include <optional>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+using namespace solidity::util;
+using namespace solidity::smtutil;
+
+void MemoryLoadResolver::run(OptimiserStepContext& _context, Block& _ast)
+{
+	MemoryLoadResolver{
+		_context.dialect,
+		SideEffectsPropagator::sideEffects(_context.dialect, CallGraphGenerator::callGraph(_ast)),
+		SSAValueTracker::ssaVariables(_ast),
+		MSizeFinder::containsMSize(_context.dialect, _ast)
+	}(_ast);
+}
+
+void MemoryLoadResolver::visit(Expression& _e)
+{
+	Scoper::visit(_e);
+
+	// We skip the replacing if there is msize. TODO do we need this, and can we do it earlier?
+	if (m_containsMsize)
+		return;
+
+	// TODO Extract this
+	// Check and replace the value of mload(x), if possible
+	if (FunctionCall const* functionCall = get_if<FunctionCall>(&_e))
+		if (functionCall->functionName.name == m_dialect.memoryLoadFunction(YulString{})->name)
+
+			if (Identifier const* identifier = get_if<Identifier>(&functionCall->arguments.front()))
+				if (m_memory.count(identifier->name))
+				{
+					auto& value = m_memory.at(identifier->name);
+					if (inScope(value))
+						_e = Identifier{locationOf(_e), value};
+				}
+}
+
+void MemoryLoadResolver::operator()(VariableDeclaration& _v)
+{
+	// Otherwise, we simply assign a free variable to it.
+	Scoper::operator()(_v);
+
+	if (
+		_v.variables.size() == 1 &&
+		_v.value &&
+		m_ssaVariables.count(_v.variables.front().name)
+	)
+	{
+		YulString& variableName = _v.variables.front().name;
+		bool const inserted = m_variables.insert({
+			variableName,
+			m_solver->newVariable("yul_" + variableName.str(), defaultSort())
+			}).second;
+		yulAssert(inserted, "");
+		std::cout << "encoding: " << variableName.str() << std::endl;
+		m_solver->addAssertion(m_variables.at(variableName) == encodeExpression(*_v.value));
+	}
+}
+
+void MemoryLoadResolver::operator()(FunctionCall& _f)
+{
+	Scoper::operator()(_f);
+
+	SideEffectsCollector sideEffects{m_dialect, _f, &m_functionSideEffects};
+
+	if (sideEffects.invalidatesMemory())
+	{
+		set<YulString> keysToErase;
+		for ([[maybe_unused]] auto const& [key, value]: m_memory)
+			if (invalidatesMemoryLocation(key, _f))
```

Comment:
This will create tons of queries...
I'm wondering if this can be somehow simplified...